rohmu/notifier/http.py

```diff
@@ -0,0 +1,93 @@
+"""Copyright (c) 2022 Aiven, Helsinki, Finland. https://aiven.io/"""
+from .interface import Notifier
+from datetime import datetime
+from enum import Enum, unique
+from queue import SimpleQueue
+from typing import Optional, Any, Dict
+
+import requests
+import threading
+
+
+@unique
+class Operation(Enum):
+    UPLOAD = "UPLOAD"
+    DELETE = "DELETE"
+    DELETE_TREE = "DELETE_TREE"
+
+
+def _get_requests_session() -> requests.Session:
+    retry = requests.adapters.Retry(
+        total=3,
+        backoff_factor=0.5,
+        status_forcelist=[429, 500, 502, 503, 504],
+        method_whitelist={"POST"},
+    )
+    adapter = requests.adapters.HTTPAdapter(max_retries=retry)
+    request_session = requests.Session()
+    request_session.mount("http://", adapter)
+    return request_session
+
+
+class HTTPNotifier(Notifier):
+    def __init__(self, url: str) -> None:
+        self._session = _get_requests_session()
+        self._url = url
+
+    def object_created(self, key: str, size: Optional[int]) -> None:
+        self._session.post(
+            self._url,
+            {
+                "op": Operation.UPLOAD.value,
+                "key": key,
+                "size": size,
+                "last_modified": datetime.now(),
+            },
+        )
+
+    def object_deleted(self, key: str) -> None:
+        self._session.post(self._url, {"op": Operation.DELETE.value, "key": key})
+
+    def tree_deleted(self, key: str) -> None:
+        self._session.post(self._url, {"op": Operation.DELETE_TREE.value, "key": key})
+
+
+@dataclass
+class _HTTPNotifyJob:
+    url: str
+    data: Dict[str, Any]
+
+
+def _background_http_request(queue: "SimpleQueue[_HTTPNotifyJob]") -> None:
+    with _get_requests_session() as session:
+        notification = queue.get()
+        session.post(notification.url, notification.data)
+
+
+class BackgroundHTTPNotifier(Notifier):
+    def __init__(self, url: str) -> None:
+        self._url = url
+        self._queue = SimpleQueue()
+        self._thread = threading.Thread(
+            target=_background_http_request,
+            args=(self._queue,),
+            name="BackgroundHTTPNotifierThread",
+            daemon=True,
```

Comment:
The publishing of the data is done on a best effort basis, so it is okay for the queue to have a few items left over when the interpreter exits.

The alternative would be to introduce synchronization for cleanly shutting down the background threads, which would introduce more changes to the codebase. If we ever want to make a stronger guarantee we can add it later.

Reply:
one failure in post request will cause all further requests to fail. A threadpool make sure all threads are cleaned when the pool object goes out of scope. currently there is a `q.get` which is blocking. so unless  something asks that `q.get` to return the thread is going to blocked

Reply:
> so unless something asks that q.get to return the thread is going to blocked

That was intentional, it is simpler this way. This instead doesn't block the python process on exit because that thread never leaves.

> one failure in post request will cause all further requests to fail.

Will add exception handling.

Reply:
> This instead doesn't block the python process on exit because that thread never leaves.

can you say that again ? do you mean **does** block ? so restarting *hoards and astacus will get blocked, unless  ?
The interpreter does get blocked.

```

In [1]: from threading import Thread

In [2]: from queue import Queue

In [3]: def wp(q):
   ...:     while True:
   ...:         o = q.get()
   ...:         print(o)
   ...:

In [4]: t = Thread(target=wp, args=(Queue(),))

In [5]: t.start()

In [6]:
Do you really want to exit ([y]/n)? y


^CException ignored in: <module 'threading' from '/usr/lib64/python3.10/threading.py'>
Traceback (most recent call last):
  File "/usr/lib64/python3.10/threading.py", line 1558, in _shutdown
    lock.acquire()
KeyboardInterrupt:
➜  aiven-core git:(13653b8e4a)
```

Reply:
Maybe I should be more explicit, the line this comment is added to is important, namely you **need** the flag `daemon=True` for the behavior I mentioned

Reply:
yup, missed the `daemon` flag entirely

Reply:
Maybe we should explicitly add the explanation as code comment?